cpp/src/hash/hash_functions.cuh

```diff
@@ -78,9 +79,31 @@ struct MurmurHash3_32
 
       return combined;
     }
-  
-    __forceinline__ 
-    __host__ __device__ result_type operator()(const Key& key) const
+    
+    template <typename TKey = Key>
+    typename std::enable_if_t<
+      std::is_same<TKey, cudf::bool8>::value,
+    result_type>
+    __forceinline__ __host__ __device__ operator()(const TKey& key) const
+    {
+      return this->operator()(static_cast<bool>(static_cast<int8_t>(key)));
+    }
+
+    template <typename TKey = Key>
+    typename std::enable_if_t<
+      std::is_same<TKey, cudf::experimental::bool8>::value,
+    result_type>
+    __forceinline__ __host__ __device__ operator()(const TKey& key) const
+    {
+      return this->operator()(static_cast<bool>(static_cast<uint8_t>(key)));
```

Comment:
Remind me why the cast to `bool` is necessary? I don't like that we're feeding a machine size dependent type into a function that is processing the bytes of the type.

Reply:
This cast clamps the value to true/false. I was surprised this is necessary since the current bool8 wrapper has [an `int8_t` conversion operator](https://github.com/rapidsai/cudf/blob/005876afb628d041e1d33cacc398cbd9723ac419/cpp/include/cudf/utilities/legacy/wrapper_types.hpp#L102) that casts to the value to bool before casting to `int8_t`, but it doesn't seem to be getting called.

Here's the commit where I removed it: https://github.com/rapidsai/cudf/pull/3087/commits/0e4a978adc37c83c61cd67c7d1381fb57a22a849, and here's the `HASHING_TEST` failure in CI: https://gpuci.gpuopenanalytics.com/blue/organizations/jenkins/rapidsai%2Fgpuci%2Fcudf%2Fprb%2Fcudf-gpu-build/detail/cudf-gpu-build/1558/pipeline#log-1625

Reply:
Is this only a problem for the old bool8 wrapper then?

Reply:
I think it's a problem for both, because both their respective conversion operators are explicit. I thought `static_cast` would invoke explicit conversion operators, but maybe not?

Reply:
they definitely should be invoking that conversion operator. This is worth looking into further.

Reply:
Figured out the problem, fix in https://github.com/rapidsai/cudf/pull/3087/commits/59e5f978db6b44bc5dc21891f17dd17b20c65b57